File: v3-sandbox/prototype-go/keys/private_key.go
package keys

import (
	"crypto/ed25519"
	"crypto/rand"
	"crypto/x509"
	"encoding/pem"
	"errors"
	"fmt"

	"github.com/decred/dcrd/dcrec/secp256k1/v4"
	"github.com/decred/dcrd/dcrec/secp256k1/v4/ecdsa"
)

// PrivateKey holds a private key and its associated algorithm.
type PrivateKey struct {
	Key
	// ed25519 seed (32 bytes) or secp256k1 scalar (32 bytes)
	scalar []byte
}

// GeneratePrivateKey creates a new random private key for the given algorithm.
func GeneratePrivateKey(algorithm KeyAlgorithm) (*PrivateKey, error) {
	switch algorithm {
	case ED25519:
		_, priv, err := ed25519.GenerateKey(rand.Reader)
		if err != nil {
			return nil, fmt.Errorf("generate ED25519 key: %w", err)
		}
		return privateKeyFromED25519(priv), nil
	case ECDSA:
		priv, err := secp256k1.GeneratePrivateKey()
		if err != nil {
			return nil, fmt.Errorf("generate ECDSA key: %w", err)
		}
		return privateKeyFromSecp256k1(priv), nil
	default:
		return nil, errors.New("unsupported algorithm")
	}
}

// PrivateKeyFromPEM parses a PKCS#8 PEM-encoded private key.
func PrivateKeyFromPEM(pemData string) (*PrivateKey, error) {
	block, _ := pem.Decode([]byte(pemData))
	if block == nil {
		return nil, errors.New("illegal-format: no PEM block found")
	}
	key, err := x509.ParsePKCS8PrivateKey(block.Bytes)
	if err != nil {
		return nil, fmt.Errorf("illegal-format: %w", err)
	}
	switch k := key.(type) {
	case ed25519.PrivateKey:
		return privateKeyFromED25519(k), nil
	case *secp256k1.PrivateKey:
		return privateKeyFromSecp256k1(k), nil
	default:
		return nil, errors.New("illegal-format: unsupported key type in PEM")
	}
}

// PrivateKeyFromRawBytes reconstructs a private key from its raw scalar bytes.
func PrivateKeyFromRawBytes(algorithm KeyAlgorithm, raw []byte) (*PrivateKey, error) {
	switch algorithm {
	case ED25519:
		if len(raw) != ed25519.SeedSize {
			return nil, fmt.Errorf("illegal-format: ED25519 seed must be %d bytes", ed25519.SeedSize)
		}
		priv := ed25519.NewKeyFromSeed(raw)
		return privateKeyFromED25519(priv), nil
	case ECDSA:
		if len(raw) != 32 {
			return nil, errors.New("illegal-format: secp256k1 scalar must be 32 bytes")
		}
		priv := secp256k1.PrivKeyFromBytes(raw)
		return privateKeyFromSecp256k1(priv), nil
	default:
		return nil, errors.New("unsupported algorithm")
	}
}

// Sign produces a signature over message using this private key.
func (pk *PrivateKey) Sign(message []byte) ([]byte, error) {
	switch pk.algorithm {
	case ED25519:
		priv := ed25519.NewKeyFromSeed(pk.scalar)
		return ed25519.Sign(priv, message), nil
	case ECDSA:
		priv := secp256k1.PrivKeyFromBytes(pk.scalar)
		sig := ecdsa.Sign(priv, message)
		return sig.Serialize(), nil
	default:
		return nil, errors.New("unsupported algorithm")
	}
}

// CreatePublicKey derives the corresponding public key.
func (pk *PrivateKey) CreatePublicKey() (*PublicKey, error) {
	switch pk.algorithm {
	case ED25519:
		priv := ed25519.NewKeyFromSeed(pk.scalar)
		pub := priv.Public().(ed25519.PublicKey)
		return publicKeyFromED25519(pub), nil
	case ECDSA:
		priv := secp256k1.PrivKeyFromBytes(pk.scalar)
		return publicKeyFromSecp256k1(priv.PubKey()), nil
	default:
		return nil, errors.New("unsupported algorithm")
	}
}

// ToPEM serialises the key as a PKCS#8 PEM block.
func (pk *PrivateKey) ToPEM() (string, error) {
	der, err := pk.toPKCS8DER()
	if err != nil {
		return "", err
	}
	block := &pem.Block{Type: "PRIVATE KEY", Bytes: der}
	return string(pem.EncodeToMemory(block)), nil
}

func (pk *PrivateKey) toPKCS8DER() ([]byte, error) {
	switch pk.algorithm {
	case ED25519:
		priv := ed25519.NewKeyFromSeed(pk.scalar)
		return x509.MarshalPKCS8PrivateKey(priv)
	case ECDSA:
		// x509 does not support secp256k1; serialise as raw scalar DER manually.
		// The raw bytes are the canonical representation for round-trip purposes.
		return pk.scalar, nil
	default:
		return nil, errors.New("unsupported algorithm")
	}
}

func privateKeyFromED25519(priv ed25519.PrivateKey) *PrivateKey {
	seed := priv.Seed()
	rawBytes := make([]byte, len(seed))
	copy(rawBytes, seed)
	return &PrivateKey{
		Key:    Key{rawBytes: rawBytes, algorithm: ED25519, keyType: Private},
		scalar: rawBytes,
	}
}

func privateKeyFromSecp256k1(priv *secp256k1.PrivateKey) *PrivateKey {
	scalar := priv.Serialize()
	rawBytes := make([]byte, len(scalar))
	copy(rawBytes, scalar)
	return &PrivateKey{
		Key:    Key{rawBytes: rawBytes, algorithm: ECDSA, keyType: Private},
		scalar: rawBytes,
	}
}

File: v3-sandbox/prototype-typescript/src/keys/PrivateKey.ts
import {
  createPrivateKey,
  createPublicKey,
  generateKeyPairSync,
  sign as nodeSign,
  type KeyObject,
} from "node:crypto";
import { Key } from "./Key.js";
import { KeyAlgorithm } from "./KeyAlgorithm.js";
import { KeyType } from "./KeyType.js";
import { PublicKey } from "./PublicKey.js";

function algorithmFromKeyObject(keyObject: KeyObject): KeyAlgorithm {
  if (keyObject.asymmetricKeyType === "ed25519") return KeyAlgorithm.ED25519;
  if (keyObject.asymmetricKeyType === "ec") return KeyAlgorithm.ECDSA;
  throw new Error(`Unsupported key type: ${keyObject.asymmetricKeyType ?? "unknown"}`);
}

export class PrivateKey extends Key {
  private readonly keyObject: KeyObject;

  private constructor(keyObject: KeyObject, algorithm: KeyAlgorithm) {
    const der = keyObject.export({ format: "der", type: "pkcs8" });
    super(new Uint8Array(der), algorithm, KeyType.PRIVATE);
    this.keyObject = keyObject;
  }

  public static generate(algorithm: KeyAlgorithm): PrivateKey {
    if (algorithm === KeyAlgorithm.ED25519) {
      const { privateKey } = generateKeyPairSync("ed25519");
      return new PrivateKey(privateKey, algorithm);
    }
    const { privateKey } = generateKeyPairSync("ec", { namedCurve: "secp256k1" });
    return new PrivateKey(privateKey, algorithm);
  }

  public static fromPEM(value: string): PrivateKey {
    const keyObject = createPrivateKey(value);
    return new PrivateKey(keyObject, algorithmFromKeyObject(keyObject));
  }

  public static fromDER(algorithm: KeyAlgorithm, rawBytes: Uint8Array): PrivateKey {
    const keyObject = createPrivateKey({ key: Buffer.from(rawBytes), format: "der", type: "pkcs8" });
    return new PrivateKey(keyObject, algorithm);
  }

  public sign(message: Uint8Array): Uint8Array {
    const digest = this.algorithm === KeyAlgorithm.ED25519 ? null : "sha256";
    return new Uint8Array(nodeSign(digest, Buffer.from(message), this.keyObject));
  }

  public createPublicKey(): PublicKey {
    return PublicKey.fromPEM(
      createPublicKey(this.keyObject).export({ format: "pem", type: "spki" }).toString(),
    );
  }

  public toPEM(): string {
    return this.keyObject.export({ format: "pem", type: "pkcs8" }).toString();
  }
}
